karoo_gp/base_class.py

```diff
@@ -14,182 +14,162 @@
+    It's composed of a heirarchal structure of classes. Below are some of the
+    more import attributes and methods for each class, and their organization:
+
+    BaseGP
+    ├─ .scoring = {field: func...}      - funcs are passed (y_true, y_pred)
+    ├─ .encoder(y)                      - Initialize with y_train
+    │   └─ .decode(pred)                - transforms prediction to match y
+    │
+    ├─ .fitness_compare(a, b)           - determines the fitter of two trees
+    │
+    ├─ .terminals = Terminals           - list of terminals + constants
+    │   └─ .get(instx)                  - returns list of terminals from instx
+    │
+    ├─ .functions = Functions           - operators and associated type/arity
+    │   └─ .get(instx)                  - returns list of functions from instx
+    │
+    ├─ .engine = Engine                 - Numpy for cpu, Tensorflow for gpu
+    │   └─ .predict(trees, X, X_hash)   - returns X predictions for each tree
+    │
+    ├─ .population = Population         - an isolated group of trees
+    │   ├─ .fittest = Tree              - return the fittest tree
+    │   ├─ .trees                       - list of Trees, length tree_pop_max
+    │   │   ├─ ...
+    │   │   └─ Tree                     - an evolvable expression tree
+    │   │      ├─ .expression           - a string of sympified expr, e.g. a*b
+    │   │      ├─ .score                - a dict of results matching `scoring`
+    │   │      ├─ .root = Branch        - a recursive node which forms a Tree
+    │   │      │   ├─ .symbol           - a terminal ('a') or function ('*')
+    │   │      │   ├─ .arity            - instructions for generating children
+    │   │      │   ├─ .parent           - the node immediately above
+    │   │      │   └─ .children         - the nodes immediately below
+    │   │      │       └─ ...
+    │   │      └─ mutate, crossover...  - methods used by population.evolve
+    │   │
+    │   ├─ .evaluate(X, y)              - predict and score trees
+    │   └─ .evolve()                    - return a new generation of trees
+    │
+    ├─ .fit(X, y)                       - evolve expressions to predict y
+    ├- .predict(X)                      - return predicted y values for X
+    └- .score(pred, y)                  - return score of prediction against y
+    """
+
+    def __init__(
+        self, tree_type='r', tree_depth_base=3, tree_depth_max=3,
+        tree_depth_min=1, tree_pop_max=100, gen_max=10, tourn_size=7,
+        filename='', output_dir='', evolve_repro=0.1, evolve_point=0.1,
+        evolve_branch=0.2, evolve_cross=0.6, display='s', precision=None,
+        swim='p', mode='s', seed=None, pause_callback=None,
+        engine_type='numpy', tf_device="/gpu:0", tf_device_log=False,
+        functions=None, terminals=None, test_size=0.2, scoring=None,
+        higher_is_better=False, _cache={}):
+        """Initialize a Karoo_GP object with given parameters"""
```

Comment:
Shouldn't this be `BaseGP`?
```suggestion
        """Initialize a BaseGP object with given parameters"""
```